idx/memory/memory.go

```diff
@@ -518,6 +518,57 @@ func (m *MemoryIdx) delete(orgId int, n *Node) ([]string, error) {
 	return deletedIds, nil
 }
 
+// delete metricDefs from the index if they have not been seen since "oldest"
+func (m *MemoryIdx) Prune(orgId int, oldest time.Time) ([]schema.MetricDefinition, error) {
+	oldestUnix := oldest.Unix()
+	pruned := make([]schema.MetricDefinition, 0)
+	pre := time.Now()
+	m.RLock()
+	defer m.RUnlock()
+	orgs := []int{orgId}
+	if orgId == -1 {
+		log.Info("pruning stale metricDefs across all orgs")
+		orgs = make([]int, len(m.Tree))
+		i := 0
+		for org := range m.Tree {
+			orgs[i] = org
+			i++
+		}
+	}
+
+	for _, org := range orgs {
+		tree, ok := m.Tree[org]
+		if !ok {
+			continue
+		}
+		for _, n := range tree.Items {
+			if !n.Leaf {
+				continue
+			}
+			for i := len(n.Children) - 1; i >= 0; i-- {
```

Comment:
If you delete from a slice while iterating over it, you are not going to get the result you expected as the positions of the items in the slice change when you delete an item.

https://play.golang.org/p/m5Pleaw2Os